lib/puppetserver/ca/generate_action.rb

```diff
@@ -0,0 +1,249 @@
+require 'optparse'
+require 'openssl'
+require 'puppetserver/utils/file_utils'
+require 'puppetserver/settings/ttl_setting'
+
+module Puppetserver
+  module Ca
+    class GenerateAction
+      CA_EXTENSIONS = [
+        ["basicConstraints", "CA:TRUE", true],
+        ["keyUsage", "keyCertSign, cRLSign", true],
+        ["subjectKeyIdentifier", "hash", false],
+        ["authorityKeyIdentifier", "keyid:always", false]
+      ]
+      DEFAULT_SIGNING_DIGEST = OpenSSL::Digest::SHA256.new
+
+      SUMMARY = "Generate a root and intermediate signing CA for Puppet Server"
+      BANNER = <<-BANNER
+Usage:
+  puppetserver ca generate [--help|--version]
+  puppetserver ca generate [--config PATH]
+
+Description:
+Generate a root and intermediate signing CA for Puppet Server
+and store generated CA keys, certs, and crls on disk.
+
+To determine the target location, the default puppet.conf
+is consulted for custom values. If using a custom puppet.conf
+provide it with the --config flag
+Options:
+BANNER
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        # Validate config_path provided
+        config_path = input['config']
+        errors = Puppetserver::Utils::FileUtilities.validate_file_paths([config_path])
+        log_errors_and_return(errors) if errors.any?
+
+        # Load, resolve, and validate puppet config settings
+        puppet = PuppetConfig.parse(config_path)
+        log_errors_and_return(puppet.errors) if puppet.errors.any?
+
+        # Generate root and intermediate ca and put all the certificates, crls,
+        # and keys where they should go.
+        generate_root_and_intermediate_ca(puppet)
+
+        # Puppet's internal CA expects these file to exist.
+        Puppetserver::Utils::FileUtilities.ensure_file(puppet.settings[:serial], "001", 0640)
```

Comment:
I think we want this to be "003", since the two other certificates have serials 1 and 2.

Reply:
Talked with Justin, and because this is what the intermediate CA will use to sign certs, it should be 1.  (Since this is what we will look at when the intermediate signs its FIRST certificate).  The root gets a 1 since it signed itself and the intermediate gets a 2 since its the second cert the root signed. Does that sound right?